### ISSUE TITLE
Quantum Operation works on single qubits, but not on elements of different arrays

### ISSUE THREAD

**Issue by marco-lewis** [eth-sri/silq#23]
I have a quantum gate, `W(a:B,b:B)`, which basically performs a bunch of `cnot` and controlled-`H` operations between the qubits. Neither of the variables are constant because they are each a target of some `cnot` operation.

The following program performs as expected:
```
x := 0:B;
y := 1:B;
[x, y] := W(x, y);
```

I would like to generalize this so that `x` and `y` are of type `uint[n]`, where I would perform the operation between pairs of qubits that share the same index. The following code, which looks like it could work, produces an error though:
```
x:=0:uint[n];
\\ Do some stuff to x
y:=0:uint[n];
\\ Do some stuff to y
for i in [0..n){
    [x[i], y[i]] := W(x[i], y[i]); \\Error: only swapping supported in permute statement
}
```

Is there some functionality to do this already or is this a feature that could be added?
I can work around this by moving the definition of `W` into the for loop and perform the operation directly, but if I need to perform `W` multiple times it would be easier to call the function.

**Comment by tgehr** [comment]
Thanks for letting us know! This is indeed a feature that can/should/will be added, but I have not gotten around to implementing it.

Meanwhile, you can use one of the following workarounds:

Option 1: Procedural approach using CNOT:
```
n := 2;
def main(){
	x:=0:uint[n];
	// do some stuff to x
	y:=0:uint[n];
	// do some stuff to y
	for i in 0..n{
		// move (x[i], y[i]) out of array:
		(a,b) := (x[i],y[i]):𝔹^2;
		x[i] := CNOT(a,x[i]);
		y[i] := CNOT(b,y[i]);
		// apply gate
		(a,b) := W(a,b);
		// move (a,b) back into array:
		x[i] := CNOT(a,x[i]);
		y[i] := CNOT(b,y[i]);
		forget((a,b)=(x[i],y[i]));
	}
	return (x,y);
}
```

Option 2: Functional approach:
```
import util;
n:=2;
def main(){
	x:=0:uint[n];
	// do some stuff to x
	y:=0:uint[n];
	// do some stuff to y
	(x,y):=unzip(map(W, zip(x as 𝔹^n,y as 𝔹^n))) as int[n]^2;
	return (x,y);
}
```
Where util.slq contains the following helper functions:
```
def map[a,b,n:!ℕ](f: a!→mfree b, xs: a^n)mfree: b^n{
	if n==0{ forget(xs=()); result := () coerce b^n; }
	else { (x,)~xs:=xs; result := (f(x),)~map(f, xs); }
	return result;
}

def zip[a,b,n:!ℕ](xs:a^n,ys:b^n)qfree:(a×b)^n{
	xs:=xs:a[];
	ys:=ys:b[];
	r:=[]:(a×b)[];
	repeat n{
		(x,)~xs:=xs;
		(y,)~ys:=ys;
		r~=[(x,y)];
	}
	return r coerce (a×b)^n;
}

def unzip[a,b,n:!ℕ](xs:(a×b)^n)qfree:a^n×b^n{
	xs:=xs:(a×b)[];
	rs:=[]:a[];
	qs:=[]:b[];
	repeat n{
		((r,q),)~xs:=xs;
		rs~=[r];
		qs~=[q];
	}
	return (rs,qs) coerce a^n×b^n;
}
```

**Comment by marco-lewis** [comment]
Good to hear, I'll use Option 1 for what I'm doing and hopefully will be able to replace it in the future.

**Comment by tgehr** [comment]
Implemented in c0029ff86575c78cb1580a14d5648ade7f7aeed4, works with vscode plugin 0.0.33.

I also pitched in some extra generality. For example, an operation can be applied to multiple qubits picked from any number of arrays, possibly multiple qubits from each array, where some qubits are consumed and replaced and others are used as controls, given that a simple alias analysis can prove that the same qubit participates in the same operation at most once.

@marco-lewis You should be able to remove the workaround now. Just let me know if you run into any further issues or if you find some case where improving the alias analysis further would be helpful.
